### PR TITLE
Add sampling to vlm pipeline by Sampler

### PR DESCRIPTION
**Pull request by sbalandi**
for review 
Task [CVS-152890](https://jira.devtools.intel.com/browse/CVS-152890)

### CODE REVIEW

**Comment by sbalandi** [comment]
@Wovchena  @ilya-lavrenov @iefode @popovaan please, review

File: src/cpp/src/vlm_pipeline.cpp [hunk 1]
```diff
@@ -296,6 +300,163 @@ ov::Tensor resample(VLMPipeline& pipe, const ov::Tensor& encoded_image, const st
 }
 }
 
+
+void forward_embedings_and_lm(SequenceGroup::CPtr sequence_group, ov::InferRequest& embedding, ov::InferRequest& language, const VLMConfig m_vlm_config, const std::shared_ptr<Sampler> sampler) {
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/cpp/src/vlm_pipeline.cpp:304]: maybe we can merge this `m_vlm_config.scale_emb` applying as model post-processing?

File: src/cpp/src/vlm_pipeline.cpp [hunk 2]
```diff
@@ -447,31 +608,16 @@ DecodedResults VLMPipeline::generate(
...
-    m_language.get_tensor("beam_idx").set_shape({ BATCH_SIZE });
-    m_language.get_tensor("beam_idx").data<int32_t>()[0] = 0;
+    
+    std::shared_ptr<Sampler> sampler = std::make_shared<Sampler>(m_tokenizer);
 
-    m_language.infer();
+    std::vector<SequenceGroup::Ptr> requests;
+    auto attention_size = m_language.get_tensor("attention_mask").get_size();	    // request_id, input_ids, generation_config, block_size, enable_prefix_caching
+    // now we have one prompt as input, so we need one request
+    SequenceGroup::Ptr sequence_group = std::make_shared<SequenceGroup>(0, encoded_input, generation_config, 1, false);
```
Comment (ilya-lavrenov) [inline src/cpp/src/vlm_pipeline.cpp:617]: @Wovchena do encoded_input and inputs_embeds have the same sequence_len size ? 
Sequence group assumes std::vector with inputs_ids of the size equal to sequence_len (prompt len at the beginning)

File: src/cpp/src/vlm_pipeline.cpp [hunk 3]
```diff
@@ -296,6 +300,163 @@ ov::Tensor resample(VLMPipeline& pipe, const ov::Tensor& encoded_image, const st
...
+        * beam_idx_data = beam_idx.data<int32_t>();
+
+    std::vector<Sequence::CPtr> running_sequences = sequence_group->get_running_sequences();
+    size_t num_running_sequences = running_sequences.size();
+    size_t num_scheduled_tokens = sequence_group->get_num_scheduled_tokens();
+    size_t group_position_id = sequence_group->get_num_processed_tokens();
+
+    // spec: In case of multiple input tokens for current sequence (prompt_len > 1),
+    // context_len corresponds to first token within subgroup of scheduled tokens
+    size_t group_context_len = group_position_id;
```
Comment (ilya-lavrenov) [inline src/cpp/src/vlm_pipeline.cpp:331]: not sued

File: src/cpp/src/sampler.cpp [hunk 4]
```diff
@@ -561,6 +577,16 @@ std::vector<int64_t> Sampler::_try_finish_generation(SequenceGroup::Ptr & sequen
 }
 
 
+std::vector<int32_t> Sampler::get_beam_idxs(uint64_t request_id) {
+    std::vector<int32_t> beams;
+    if (m_beam_search_info.find(request_id) != m_beam_search_info.end()) {
+        GroupBeamSearcher beam_searcher = m_beam_search_info.at(request_id);
+        std::vector<int32_t> beams = beam_searcher.get_beam_idxs();
+    }
+    return beams;
```
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/cpp/src/sampler.cpp:586]: maybe we can throw exception? 
or return { 0 }  which can be used even on prompt phase.

File: src/cpp/src/vlm_pipeline.cpp (hunk 2 above)
Comment (Wovchena) [reply src/cpp/src/vlm_pipeline.cpp:617]: Yes

File: src/cpp/src/vlm_pipeline.cpp [hunk 5]
```diff
@@ -447,31 +608,16 @@ DecodedResults VLMPipeline::generate(
...
+    
+    std::shared_ptr<Sampler> sampler = std::make_shared<Sampler>(m_tokenizer);
 
-    m_language.infer();
+    std::vector<SequenceGroup::Ptr> requests;
+    auto attention_size = m_language.get_tensor("attention_mask").get_size();	    // request_id, input_ids, generation_config, block_size, enable_prefix_caching
+    // now we have one prompt as input, so we need one request
+    SequenceGroup::Ptr sequence_group = std::make_shared<SequenceGroup>(0, encoded_input, generation_config, 1, false);
+    sequence_group->set_sequence_group_ptr(sequence_group);
+    requests.push_back(sequence_group);
```
Comment (ilya-lavrenov) [inline src/cpp/src/vlm_pipeline.cpp:620]: could you please also drop VLM specific sampling https://github.com/openvinotoolkit/openvino.genai/blob/master/src/cpp/src/vlm_sampling.hpp ?

**Comment by ilya-lavrenov** [comment]
@sbalandi could you please apply comments and submit PR to GenAI repo?

File: src/cpp/src/vlm_pipeline.cpp (hunk 5 above)
Comment (sbalandi) [reply src/cpp/src/vlm_pipeline.cpp:531]: removed

File: src/cpp/src/sampler.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilya-lavrenov) [reply src/cpp/src/sampler.cpp]: at the beginning, when there is only once sequence, we need { 0 }

on second and other iterations, `m_beam_search_info` is here and it provides the proper value of `get_beam_idxs`

**Comment by sbalandi** [comment]
moved to pr to GenAI repo https://github.com/openvinotoolkit/openvino.genai/pull/950

File: src/cpp/src/vlm_pipeline.cpp (hunk 2 above)
Comment (sbalandi) [reply src/cpp/src/vlm_pipeline.cpp]: history have added

File: src/cpp/src/sampler.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbalandi) [reply src/cpp/src/sampler.cpp]: done

**Pull request closed without merging**